### PR TITLE
Markers dock

### PR DESCRIPTION
**Pull request by bmatherly**
Create a new dock panel to list all markers, allow sorting and editing. Clicking on a marker selects it and seeks the timeline to that marker.

image

Interested in comments on usability

### CODE REVIEW

**Review by ddennedy — CHANGES_REQUESTED** [review]
I am still reviewing, but some initial feedback. There should be a shortcut for this. I suggest Ctrl+Shift+6 (because it continues the Ctrl+number theme and on US keyboard, carot is above 6). You can also add to the end of the main toolbar. I usually turn off text under icons and have plenty more room. Besides, it can overflow. I think we will leave this out of any of the preset layouts for now since that is a pain in the butt to maintain. I do like the default placement when I open it in the Editing layout. I did also test some functionality, and it is working good.

File: src/widgets/editmarkerwidget.cpp [hunk 1]
```diff
@@ -111,22 +134,29 @@ void EditMarkerWidget::on_colorButton_clicked()
...
 
 void EditMarkerWidget::updateDuration()
 {
-    int duration = m_endSpinner->value() - m_startSpinner->value() + 1;
-    m_durationLabel->setText(MLT.producer()->frames_to_time(duration));
+    if (MLT.producer()) {
+        int duration = m_endSpinner->value() - m_startSpinner->value() + 1;
+        m_durationLabel->setText(MLT.producer()->frames_to_time(duration));
+    } else {
+        m_durationLabel->setText("00:00:00:00");
```
[the remark this answers is not in the document]
Comment (bmatherly) [reply src/widgets/editmarkerwidget.cpp:160]: I would like to, but this is the "else" case of "if (MLT.producer())". So I would need some other object to call frames_to_time() on. Maybe I could just clear out the duration field to be blank.

**Comment by bmatherly** [comment]
> I suggest Ctrl+Shift+6 (because it continues the Ctrl+number theme and on US keyboard, caret is above 6).

Done

> You can also add to the end of the main toolbar. I usually turn off text under icons and have plenty more room. 

Done

File: src/widgets/editmarkerwidget.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (ddennedy) [reply src/widgets/editmarkerwidget.cpp:160]: Oh yeah, that makes sense. But I could not figure out how to make this appear. When I close a project with markers and one of the markers selected in the panel, it still shows the old info in the edit form within the panel. I think that is the last thing that needs to be addressed here.

[the remark this answers is not in the document]
Comment (ddennedy) [reply src/widgets/editmarkerwidget.cpp:160]: In the player, I use "--:--:--:--" to represent cleared or no timecode. Some languages use different numeric symbols than "0".

**Comment by bmatherly** [comment]
Sorry to pile on, but I added two additional features that I had been working on:

**Recent color selection in marker context menu:**

image

**New menu in Marker panel to provide "Remove All" and Column hiding:**

image

**Comment by ddennedy** [comment]
> Sorry to pile on

That's fine. I think the main thing people will miss here is a search box.

> Recent color selection in marker context menu

I think we can make that menu more narrow. Not too narrow tho because a translation for "Other" could be slighter longer than the English. But I think about half the current width is fine.

**Comment by ddennedy** [comment]
I think this looks a little nicer
```
--- src/qml/modules/Shotcut/Controls/Marker.qml
+++ src/qml/modules/Shotcut/Controls/Marker.qml
@@ -82,10 +82,11 @@ Item {
         Menu {
             id: colorMenu
             title: qsTr('Color')
+            width: 100
             Instantiator {
                 model: markers.recentColors
                 MenuItem {
-                    implicitHeight: 14
+                    implicitHeight: otherMenuItem.implicitHeight
                     id: menuItem
                     background: Rectangle {
                         color: modelData
@@ -99,6 +100,7 @@ Item {
             }
             MenuSeparator { }
             MenuItem {
+                id: otherMenuItem
                 text: qsTr('Other...')
                 onTriggered: {
                     colorDialog.color = root.markerColor
```

**Comment by bmatherly** [comment]
I took your suggestion on the color context menu size.

I also added a search box for the marker dock and converted the buttons to toolbar style to be consistent with other docks.

image

**Review by ddennedy — APPROVED** [review]

**Pull request merged**